modules/basis_universal/register_types.cpp

```diff
@@ -228,35 +234,29 @@ static Ref<Image> basis_universal_unpacker_ptr(const uint8_t *p_data, int p_size
 
 	ERR_FAIL_COND_V(!tr.validate_header(ptr, size), image);
 
+	tr.start_transcoding(ptr, size);
+
 	basist::basisu_image_info info;
 	tr.get_image_info(ptr, size, info, 0);
-
-	int block_size = basist::basis_get_bytes_per_block_or_pixel(format);
 	Vector<uint8_t> gpudata;
-	gpudata.resize(info.m_total_blocks * block_size);
-
-	{
-		uint8_t *w = gpudata.ptrw();
-		uint8_t *dst = w;
-		for (int i = 0; i < gpudata.size(); i++) {
-			dst[i] = 0x00;
-		}
+	gpudata.resize(Image::get_image_data_size(info.m_width, info.m_height, imgfmt, info.m_total_levels > 1));
 
-		int ofs = 0;
-		tr.start_transcoding(ptr, size);
-		for (uint32_t i = 0; i < info.m_total_levels; i++) {
-			basist::basisu_image_level_info level;
-			tr.get_image_level_info(ptr, size, level, 0, i);
-
-			bool ret = tr.transcode_image_level(ptr, size, 0, i, dst + ofs, level.m_total_blocks - i, format);
-			if (!ret) {
-				printf("failed! on level %u\n", i);
-				break;
-			};
-
-			ofs += level.m_total_blocks * block_size;
+	uint8_t *w = gpudata.ptrw();
+	uint8_t *dst = w;
+	for (int i = 0; i < gpudata.size(); i++) {
+		dst[i] = 0x00;
+	}
+	uint32_t mip_count = Image::get_image_required_mipmaps(info.m_orig_width, info.m_orig_height, imgfmt);
+	for (uint32_t level_i = 0; level_i < mip_count; level_i++) {
+		basist::basisu_image_level_info level;
+		tr.get_image_level_info(ptr, size, level, 0, level_i);
+		int ofs = Image::get_image_mipmap_offset(info.m_width, info.m_height, imgfmt, level_i);
+		bool ret = tr.transcode_image_level(ptr, size, 0, level_i, dst + ofs, level.m_total_blocks, format);
```

Comment:
Might be a good idea to add some checks that `Image::get_image_mipmap_size` matches the decoded size of this basisu level (something from `info`, `struct basisu_image_level_info` defined in `thirdparty/basis_universal/transcoder/basisu_transcoder.h` )

Note that at some mipmap levels, width and height might not be a multiple of 4, but the data will still be a multiple of blocks.

You could imagine potentially decoding a ktx file made with different mipmap levels, since (I believe) there is no requirement that all mipmap levels exist, and a malicious or incompatible ktx file might crash godot in such a case.